### PR TITLE
Placeholder for transitions reset to prevent code duplication for #29862

### PR DESCRIPTION
**Pull request by ffoodd**
As mentionned in #29862, resetting globally animations and transitions is too risky — but we can still reduce redundancy by using a placeholder.

It also allow resetting transitions without using the `transition()` mixin, by simply extending the placeholder.

And it's far from useless, comparing the results of `npm run bundlesize` locally:

| File | master | Pull Request |
| --- | --- | --- |
| `bootstrap.css` | 22.47 kb | 22.23 kb |
| `bootstrap.min.css` | 20.93 kb | 20.71 kb |

I'd be pleased to dicuss this, and thanks @MartijnCuppens and @patrickhlauke for your comments — I'll update my fork accordingly :)

### CODE REVIEW

**Comment by MartijnCuppens** [comment]
This looks promising: https://www.diffchecker.com/8K7T2qdj

However, there's an issue with this block:

```css
	@media (prefers-reduced-motion: reduce) {
  .form-control, .form-range::-webkit-slider-thumb, .form-range::-moz-range-thumb, .form-range::-ms-thumb, .btn, .fade, .collapsing, .progress-bar, .modal.fade .modal-dialog, .carousel-item, .carousel-fade .active.carousel-item-left,
  .carousel-fade .active.carousel-item-right, .carousel-control-prev,
  .carousel-control-next, .carousel-indicators li {
    -webkit-transition: none !important;
    -moz-transition: none !important;
    -ms-transition: none !important;
    transition: none !important;
  }
}
```

Once a browser comes across an invalid selector (`.form-range::-webkit-slider-thumb`, `.form-range::-moz-range-thumb` or `.form-range::-ms-thumb`), it will invalidate the whole selector, thus these properties will never be applied.

Not sure how we best handle this, any ideas?

**Comment by ffoodd** [comment]
You're right!

I only see two ways here:

1. managing `.form-range` transitions another way: maybe introducing another mixin that doesn't extend `%no-transition`? Or do not use the `transition()` mixin for those special cases — but I'm not convinced at all, preferring the second mixin;
2. find a way in `transition()` mixin to check if we wan't to extend `%no-transition` or not. This is probably harder, but if reachable would grant compatibility with existing codebases.

I used to have a quarantine mode especially for this invalid selector stack issue in [a11y.css](https://github.com/ffoodd/a11y.css) but it requires a boolean optionnal argument fir the mixin, which is impossible in `transition()` since it has a single `arglist` argument…

But it could be doable with selector list introspection in the mixin, as explained in [this StackOverflow answer to a question about parent selector within a mixin](https://stackoverflow.com/a/46126037).

Working on it.

**Comment by ffoodd** [comment]
Hi @MartijnCuppens, problem solved!

I managed to work around:
1. new `$quarantine-selectors` variable, `list` type containing vendor prefixes
2. new boolean variable in `transition()`: `$is-infected`, set to `false`
3. a loop over `$quarantine-selectors` to check if parent selector `&` has a part of it in quarantine  — and if so, set `$is-infected` to `true`
5. depending on `$is-infected`, print transition reset directly or extend `%no-transition`.

There might be edge cases I didn't think about, though. And it obviously need a review, please tell if variable names is inappropriate. I think they're both appropriate and fun but I may be wrong :D

FYI, it compiled with every Sass compiler except Sass 3.3 — tested on SassMeister.

File: scss/_variables.scss
Comment (MartijnCuppens) [inline scss/_variables.scss]: I would spread this over multiple lines.

**Comment by MartijnCuppens** [comment]
> There might be edge cases I didn't think about, though.

I don't think so, but I'll do some more in depth testing by the end of the week to double check this.

> And it obviously need a review, please tell if variable names is inappropriate. I think they're both appropriate and fun but I may be wrong :D

I don't mind having fun variable names, the only concern I have is that this variable might be used in the future for other purposes in which "infected" doesn't make sense. So I think it'll be better to rename it to `$pseudo-vendor-prefixes` or something alike.

File: site/assets/scss/_anchor.scss
Comment (MartijnCuppens) [inline site/assets/scss/_anchor.scss]: I would move this to `docs.scss`, `_anchor.scss` is quite a weird place to have this.

File: site/assets/scss/_anchor.scss
Comment (MartijnCuppens) [inline site/assets/scss/_anchor.scss]: `!important` needs to be added for consistency

File: site/assets/scss/_anchor.scss
Comment (ffoodd) [reply site/assets/scss/_anchor.scss]: Just did. Looks weird to me, don't you think we should add `_placeholder.scss` or something?

**Review by MartijnCuppens — APPROVED** [review]
I just fixed Dart Sass compatibility. LGTM now, but waiting for another approval from @twbs/css-review

File: scss/_variables.scss
Comment (XhmikosR) [inline scss/_variables.scss]: I'm not sure this is self-explanatory...

**Comment by XhmikosR** [comment]
Also do note @patrickhlauke's comment https://github.com/twbs/bootstrap/issues/29862#issuecomment-566526017

**Comment by ffoodd** [comment]
@XhmikosR, about @patrickhlauke I think he was concerned about my first suggestion using a universal selector to reset **every** transitions and animations.

This PRs does not, and everything is still dependant of the `transition()` mixin, so it keeps specific selectors to prevent bleeding out to author styles and allows extending and removing on a component basis.

This PR simply groups `prefers-reduced-motion` queries when possible.


—

I'm still trying to find a better comment to explain what this map is used for.

File: scss/_variables.scss
Comment (ffoodd) [inline scss/_variables.scss]: @XhmikosR This is more precise, inspired by the above `$escaped-characters` map's comment.

LGTY?

**Comment by ffoodd** [comment]
Failing check is Karma timing out… Can someone run checks again?

**Comment by MartijnCuppens** [comment]
Ping @twbs/css-review / @XhmikosR for additional review

**Comment by XhmikosR** [comment]
I still want @mdo to review this before merging.

**Comment by XhmikosR** [comment]
This needs a rebase and mdo's review

**Review by ysds — DISMISSED** [review]
LGTM.

BTW, [IE and EdgeHTML does not support the `prefers-reduced-motion`](https://caniuse.com/#search=prefers-reduced-motion), so the following CSS may do nothing:
```scss
@media (prefers-reduced-motion: reduce) {	
  .form-range::-ms-thumb {	
    -ms-transition: none;	
    transition: none;	
  }	
}
```

**Comment by ysds** [comment]
I marked this "approved" but I found a concern. Using Bootstrap SCSS in a module may cause problems.

my-component.scss
```scss
@import "~bootstrap/scss/functions";
@import "~bootstrap/scss/variable";
@import "~bootstrap/scss/mixin";

.element {
  color: $primary;
  @include transition($transition-fade); // Compile error
}
```

This is an edge case.

**Comment by ffoodd** [comment]
@ysds Indeed, because of `%no-transition` being declared in `_reboot.scss`.

The only solution I can think of is to declare this placeholder in `scss/mixins/_transition.scss`. But I find this confusing, isn't it?

For the `-ms-` thing, I guess we could prevent this by adding a secondary condition in the loop if `$selector` contains `-ms-`. It might be overkill though.

@MartijnCuppens have you any opinion on this?

**Comment by MartijnCuppens** [comment]
I don't think we should worry about the edge case. About `::-ms-thumb`, I don't think it's worth overcomplicating things to prevent the generation of these lines.

**Comment by ysds** [comment]
OK, but if user the new transition mixin the same as v4, a compile error occur in some case, so we should add a description in the migration.md.

**Comment by ysds** [comment]
Alternatively, it is OK to add an alias mixin that does not use the placeholder selector and guide them to use it.

**Comment by ffoodd** [comment]
Just a note about `::ms-thumb`, it was already generated so I guess this is not an issue at all.

@ysds I'm not sure about your alias suggestion, since it's a new major release things could break — but should be documented, obviously. I think you're right about `migration.md`, I'll try to add this today.

IMHO people able to import Bootstrap's partials and using it as modules should be aware that major versions introduce breaking changes.

**Review by ysds — APPROVED** [review]

**Comment by XhmikosR** [comment]
FYI this results in worse gzip compression ~170 bytes for bootstrap.min.css.

**Comment by MartijnCuppens** [comment]
Got different results here (`21258 bytes` before vs `21308 bytes` after), but this does seem to influence the file size in a bad way.

Might be worth reverting this then.

**Comment by XhmikosR** [comment]
[Before](https://bundlesize-store.now.sh/build?info=%7B%22files%22%3A%5B%7B%22maxSize%22%3A6656%2C%22path%22%3A%22.%2Fdist%2Fcss%2Fbootstrap-grid.css%22%2C%22size%22%3A6578%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A6144%2C%22path%22%3A%22.%2Fdist%2Fcss%2Fbootstrap-grid.min.css%22%2C%22size%22%3A5809%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A2048%2C%22path%22%3A%22.%2Fdist%2Fcss%2Fbootstrap-reboot.css%22%2C%22size%22%3A1884%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A2048%2C%22path%22%3A%22.%2Fdist%2Fcss%2Fbootstrap-reboot.min.css%22%2C%22size%22%3A1761%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A8192%2C%22path%22%3A%22.%2Fdist%2Fcss%2Fbootstrap-utilities.css%22%2C%22size%22%3A6667%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A7168%2C%22path%22%3A%22.%2Fdist%2Fcss%2Fbootstrap-utilities.min.css%22%2C%22size%22%3A6046%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A25600%2C%22path%22%3A%22.%2Fdist%2Fcss%2Fbootstrap.css%22%2C%22size%22%3A22806%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A23552%2C%22path%22%3A%22.%2Fdist%2Fcss%2Fbootstrap.min.css%22%2C%22size%22%3A21258%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A52224%2C%22path%22%3A%22.%2Fdist%2Fjs%2Fbootstrap.bundle.js%22%2C%22size%22%3A51651%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A24576%2C%22path%22%3A%22.%2Fdist%2Fjs%2Fbootstrap.bundle.min.js%22%2C%22size%22%3A24140%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A28672%2C%22path%22%3A%22.%2Fdist%2Fjs%2Fbootstrap.esm.js%22%2C%22size%22%3A28038%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A19456%2C%22path%22%3A%22.%2Fdist%2Fjs%2Fbootstrap.esm.min.js%22%2C%22size%22%3A19206%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A29696%2C%22path%22%3A%22.%2Fdist%2Fjs%2Fbootstrap.js%22%2C%22size%22%3A28671%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A17920%2C%22path%22%3A%22.%2Fdist%2Fjs%2Fbootstrap.min.js%22%2C%22size%22%3A17400%2C%22compression%22%3A%22gzip%22%7D%5D%2C%22repo%22%3A%22twbs%2Fbootstrap%22%2C%22branch%22%3A%22master%22%2C%22commit_message%22%3A%22%22%2C%22sha%22%3A%226933bd4a1d9b3b1de109fc43d9a0da858c205fb0%22%7D)

[After](https://bundlesize-store.now.sh/build?info=%7B%22files%22%3A%5B%7B%22maxSize%22%3A6656%2C%22path%22%3A%22.%2Fdist%2Fcss%2Fbootstrap-grid.css%22%2C%22size%22%3A6578%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A6144%2C%22path%22%3A%22.%2Fdist%2Fcss%2Fbootstrap-grid.min.css%22%2C%22size%22%3A5809%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A2048%2C%22path%22%3A%22.%2Fdist%2Fcss%2Fbootstrap-reboot.css%22%2C%22size%22%3A1884%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A2048%2C%22path%22%3A%22.%2Fdist%2Fcss%2Fbootstrap-reboot.min.css%22%2C%22size%22%3A1761%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A8192%2C%22path%22%3A%22.%2Fdist%2Fcss%2Fbootstrap-utilities.css%22%2C%22size%22%3A6667%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A7168%2C%22path%22%3A%22.%2Fdist%2Fcss%2Fbootstrap-utilities.min.css%22%2C%22size%22%3A6046%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A25600%2C%22path%22%3A%22.%2Fdist%2Fcss%2Fbootstrap.css%22%2C%22size%22%3A22862%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A23552%2C%22path%22%3A%22.%2Fdist%2Fcss%2Fbootstrap.min.css%22%2C%22size%22%3A21308%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A52224%2C%22path%22%3A%22.%2Fdist%2Fjs%2Fbootstrap.bundle.js%22%2C%22size%22%3A51651%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A24576%2C%22path%22%3A%22.%2Fdist%2Fjs%2Fbootstrap.bundle.min.js%22%2C%22size%22%3A24140%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A28672%2C%22path%22%3A%22.%2Fdist%2Fjs%2Fbootstrap.esm.js%22%2C%22size%22%3A28038%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A19456%2C%22path%22%3A%22.%2Fdist%2Fjs%2Fbootstrap.esm.min.js%22%2C%22size%22%3A19206%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A29696%2C%22path%22%3A%22.%2Fdist%2Fjs%2Fbootstrap.js%22%2C%22size%22%3A28671%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A17920%2C%22path%22%3A%22.%2Fdist%2Fjs%2Fbootstrap.min.js%22%2C%22size%22%3A17400%2C%22compression%22%3A%22gzip%22%7D%5D%2C%22repo%22%3A%22twbs%2Fbootstrap%22%2C%22branch%22%3A%22master%22%2C%22commit_message%22%3A%22%22%2C%22sha%22%3A%22cd7e5d111102e4422858aa6f255dc243b4cca1ca%22%7D)

**Comment by ffoodd** [comment]
Agreed, many things changed since my first draft. It's meant to decrease filesize, so it doesn't seem to work anymore :/

**Comment by ffoodd** [comment]
FWIW, thinking about that I think the increased size is kind of logical: if the selector stack extending `%no-transition` is longer than the media-query itself, it results in a longer version.

There's a way to fix this by checking selector's length before extending, but it seems over-engineered. @MartijnCuppens do you think it could worth it?

If not, I'll close #29862 :)

**Pull request merged**
